### ISSUE TITLE
Write a guide for porting trio libraries to AnyIO

### ISSUE THREAD

**Issue by agronholm** [agronholm/anyio#245]
This should hopefully speed up AnyIO's adoption if authors of existing trio libraries find it useful.

**Comment by uSpike** [comment]
Should the same be done for asyncio?  There's many more asyncio-based libraries that could be converted.  I'd be happy to contribute.

**Comment by smurfix** [comment]
Short answer: no.

Long answer: asyncio libraries ten to be an unstructured hodgepodge of callbacks (just like asyncio itself … SCNR). We don't want to replicate that.

IME it's easier and cleaner, in most cases, to convert a standard sync library to anyio. Just sprinkle with async/await, possibly add a mux/demux step if it's a protocol client, done.

**Comment by uSpike** [comment]
I'm not sure I agree with the idea that asyncio-based libraries aren't worth the effort to port to anyio.  I think there's a lot of async libraries that don't use a whole lot of asyncio and porting would be relatively easy, so long as there's not heavy use of callbacks and protocols.

Some simple stuff like "use streams instead of queues" with some examples of before/after would have been useful to me, as someone coming from asyncio-land.  Same with "use a task group instead of ensure_future() or create_task()".

**Comment by agronholm** [comment]
I agree that there is merit in providing a similar guide for asyncio.

**Comment by smurfix** [comment]
Well, I do agree that we should have a "basic asyncio patterns > equivalent anyio patterns" guide. And yes some libraries are trivially ported.

However, as soon as things become nontrivial (e.g. `asyncio.Future` is involved, callbacks doing actual work are invoked, etc.) my experience is that asyncio introduces anti-patterns which don't mesh at all with SC concepts, and the end result may well be translating a sync library is actually the cleaner *and* easier approach.

**Comment by agronholm** [comment]
I've opened #255 for the asyncio version.
